OMC does not ignore the '.swp' files

Just noticed that if someone is opening one of the `yaml` files using vim or vi, this will create a `.swp` file and omc will try to open it:
```
[vlours@supportshell-1 ~]$ oc get machineset -n openshift-machine-api
Error when trying to unmarshal file: /home/bne/vlours/03920836/0010-3920836.tar.gz/3920836/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-1d0190b2e1110469c9ea981f24f758dccacf621c5d46ab01ba1edaf5755b2ec1/namespaces/openshift-machine-api/machine.openshift.io/machinesets/.int-np-lmjcl-worker-0.yaml.swp
```
```
[vlours@supportshell-1 ~]$ ls -la /home/bne/vlours/03920836/0010-3920836.tar.gz/3920836/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-1d0190b2e1110469c9ea981f24f758dccacf621c5d46ab01ba1edaf5755b2ec1/namespaces/openshift-machine-api/machine.openshift.io/machinesets/
total 20
drwxrwxrwx+ 2 yank    yank       91 Sep  5 05:59 .
drwxrwxrwx+ 5 yank    yank       84 Sep  5 02:14 ..
-rw-rw-rw-+ 1 yank    yank     3609 Sep  5 02:05 int-np-lmjcl-worker-0.yaml
-rw-r--r--+ 1 egeorge egeorge 16384 Sep  5 05:59 .int-np-lmjcl-worker-0.yaml.swp       <<<<< 
```

In addition If you create a empty file in a folder the command will simply crash:
```
[vlours@supportshell-1 ~]$ oc get mcp
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-cd754fa9873eb664b22c894f530fdecb   True      False      False      3              3                   3                     0                      58d
worker   rendered-worker-18f2b48afb85af4e68a3b0b0afd0d208   True      False      False      6              6                   6                     0                      58d
```
```
[vlours@supportshell-1 ~]$ touch 03920836/0010-3920836.tar.gz/3920836/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-1d0190b2e1110469c9ea981f24f758dccacf621c5d46ab01ba1edaf5755b2ec1/cluster-scoped-resources/machineconfiguration.openshift.io/machineconfigpools/infra.yaml
```
```
[vlours@supportshell-1 ~]$ oc get mcp
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x198 pc=0x2ccd9fa]

goroutine 1 [running]:
github.com/gmeghnag/omc/pkg/tablegenerator.GenerateCustomResourceTable({0x0})
	/builddir/build/BUILD/omc-3.7.2/pkg/tablegenerator/tablegenerator.go:169 +0xb5a
github.com/gmeghnag/omc/cmd/get.handleObject({0x0})
	/builddir/build/BUILD/omc-3.7.2/cmd/get/get.go:443 +0x118e
github.com/gmeghnag/omc/cmd/get.getClusterScopedResources({0xc004023530, 0x12}, {0xc004412b40, 0x21}, 0xc00155f3b0)
	/builddir/build/BUILD/omc-3.7.2/cmd/get/get.go:362 +0x5c5
github.com/gmeghnag/omc/cmd/get.init.func1(0xc00019f200?, {0xc000a944c0?, 0x4?, 0x386f378?})
	/builddir/build/BUILD/omc-3.7.2/cmd/get/get.go:109 +0x4db
github.com/spf13/cobra.(*Command).execute(0x5687840, {0xc000a94480, 0x1, 0x1})
	/builddir/build/BUILD/omc-3.7.2/vendor/github.com/spf13/cobra/command.go:944 +0x867
github.com/spf13/cobra.(*Command).ExecuteC(0x5684d20)
	/builddir/build/BUILD/omc-3.7.2/vendor/github.com/spf13/cobra/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
	/builddir/build/BUILD/omc-3.7.2/vendor/github.com/spf13/cobra/command.go:992
main.main()
	/builddir/build/BUILD/omc-3.7.2/main.go:24 +0x1a
```

So it will be great to ignore empty (size = 0) and swp binary files  (.*.swp)

Thanks

--- a/cmd/get/get.go
+++ b/cmd/get/get.go
@@ -208,7 +208,10 @@ func getNamespacedResources(resourceNamePlural string, resourceGroup string, res
 			resourceDir := fmt.Sprintf("%s/namespaces/%s/%s/%s", vars.MustGatherRootPath, namespace, resourceGroup, resourceNamePlural)
 			_, err = os.Stat(resourceDir)
 			if err == nil {
-				resourcesFiles, _ := ioutil.ReadDir(resourceDir)
+				resourcesFiles, rErr := ReadDirForResources(resourceDir)
+				if rErr != nil {
+					klog.V(3).ErrorS(err, "Failed to read resources:")
+				}
 				for _, f := range resourcesFiles {
 					resourceYamlPath := resourceDir + "/" + f.Name()
 					_file, _ := ioutil.ReadFile(resourceYamlPath)
@@ -233,12 +236,11 @@ func getNamespacedResources(resourceNamePlural string, resourceGroup string, res
 			if resourceNamePlural == "pods" {
 				// tranverse the pods directory and fill in UnstructuredItems.Items
 				podsDir := fmt.Sprintf("%s/namespaces/%s/pods", vars.MustGatherRootPath, namespace)
-				pods, _ := os.ReadDir(podsDir)
+				pods, rErr := ReadDirForResources(podsDir)
+				if rErr != nil {
+					klog.V(3).ErrorS(err, "Failed to read resources:")
+				}
 				for _, pod := range pods {
-					// skip dot-prefixed files (e.g. "AppleDouble encoded Macintosh file")
-					if pod.Name()[0] == '.' {
-						continue
-					}
 					podName := pod.Name()
 					podPath := fmt.Sprintf("%s/%s/%s.yaml", podsDir, podName, podName)
 					_file, err := ioutil.ReadFile(podPath)
@@ -310,7 +312,10 @@ func getClusterScopedResources(resourceNamePlural string, resourceGroup string, 
 	_file, err := ioutil.ReadFile(resourcePath)
 	if err != nil {
 		resourceDir := fmt.Sprintf("%s/cluster-scoped-resources/%s/%s", vars.MustGatherRootPath, resourceGroup, resourceNamePlural)
-		resourcesFiles, _ := ioutil.ReadDir(resourceDir)
+		resourcesFiles, rErr := ReadDirForResources(resourceDir)
+		if rErr != nil {
+			klog.V(3).ErrorS(err, "Failed to read resources:")
+		}
 		for _, f := range resourcesFiles {
 			resourceYamlPath := resourceDir + "/" + f.Name()
 			_file, _ := ioutil.ReadFile(resourceYamlPath)

--- a/cmd/get/helpers.go
+++ b/cmd/get/helpers.go
@@ -2,12 +2,15 @@ package get
 
 import (
 	"fmt"
+	"io/fs"
 	"io/ioutil"
 	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/gmeghnag/omc/vars"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/klog/v2"
 	"sigs.k8s.io/yaml"
 )
@@ -144,7 +147,10 @@ func kindGroupNamespacedFromCrds(alias string) (string, string, bool, error) {
 	crdsPath := vars.MustGatherRootPath + "/cluster-scoped-resources/apiextensions.k8s.io/customresourcedefinitions/"
 	_, err := Exists(crdsPath)
 	if err == nil {
-		crds, _ := ioutil.ReadDir(crdsPath)
+		crds, rErr := ReadDirForResources(crdsPath)
+		if rErr != nil {
+			fmt.Fprintln(os.Stderr, rErr)
+		}
 		for _, f := range crds {
 			crdYamlPath := crdsPath + f.Name()
 			crdByte, _ := ioutil.ReadFile(crdYamlPath)
@@ -188,7 +194,10 @@ func kindGroupNamespacedFromCrds(alias string) (string, string, bool, error) {
 	}
 	home, _ := os.UserHomeDir()
 	omcCrdsPath := home + "/.omc/customresourcedefinitions/"
-	crds, _ := ioutil.ReadDir(omcCrdsPath)
+	crds, rErr := ReadDirForResources(omcCrdsPath)
+	if rErr != nil {
+		fmt.Fprintln(os.Stderr, rErr)
+	}
 	for _, f := range crds {
 		crdYamlPath := omcCrdsPath + f.Name()
 		crdByte, _ := ioutil.ReadFile(crdYamlPath)
@@ -250,4 +259,35 @@ func Exists(path string) (bool, error) {
 		return false, nil
 	}
 	return false, err
+}
+
+func ReadDirForResources(path string) ([]os.DirEntry, error) {
+	klog.V(5).Info("INFO ", fmt.Sprintf("opening '%s'\n", path))
+	return readDirForResources(os.DirFS(path))
+}
+
+// readdir wraps around fs.ReadDir and only return valid resource yaml files
+func readDirForResources(in fs.FS) ([]os.DirEntry, error) {
+	resources := make([]os.DirEntry, 0)
+	files, err := fs.ReadDir(in, ".")
+	if err == nil {
+		for _, file := range files {
+			fileName := file.Name()
+			// validate filename as per k8s validation of a resource
+			if len(validation.IsDNS1123Subdomain(fileName)) == 0 {
+				// only dirs or yaml files are expected as valid resources, e.g.:
+				//  router-default-abcde12345-fgh678 or rendered-worker-abcdef123456.yaml
+				if filepath.Ext(fileName) == ".yaml" || file.IsDir() {
+					fInfo, _ := file.Info()
+					// ignore empty files
+					if fInfo.Size() > 0 {
+						resources = append(resources, file)
+					}
+				}
+			}
+		}
+	} else {
+		err = fmt.Errorf("failed to read dir: %s", err)
+	}
+	return resources, err
 }

--- a/cmd/get/helpers_test.go
+++ b/cmd/get/helpers_test.go
@@ -1,0 +1,86 @@
+package get
+
+import (
+	"io/fs"
+	"reflect"
+	"sort"
+	"testing"
+	"testing/fstest"
+)
+
+func TestReadDirForResources(t *testing.T) {
+	tests := []struct {
+		name     string
+		in       fstest.MapFS
+		expected []string
+	}{
+		{
+			name: "read correct resource files/dirs",
+			in: fstest.MapFS{
+				"resource-file-1.yaml":            {Data: []byte("abc")},
+				"resource.yaml":                   {Data: []byte("abc")},
+				"1.yaml":                          {Data: []byte("abc")},
+				"resource.with.dot.filename.yaml": {Data: []byte("abc")},
+				"resource-directory-name":         {Data: []byte("abc"), Mode: fs.ModeDir},
+				"resource.directory.with.dot":     {Data: []byte("abc"), Mode: fs.ModeDir},
+			},
+			expected: []string{
+				"resource-file-1.yaml",
+				"resource.yaml",
+				"1.yaml",
+				"resource.with.dot.filename.yaml",
+				"resource-directory-name",
+				"resource.directory.with.dot",
+			},
+		},
+		{
+			name: "read only resource files/dirs matching the expected name convention",
+			in: fstest.MapFS{
+				"resource-file-1.yaml":             {Data: []byte("abc")},
+				"._faulthy-resource-filename.yaml": {Data: []byte("abc")}, // e.g. AppleDouble encoded Macintosh file
+				".resource-filename.yaml.swp":      {Data: []byte("abc")},
+				"-resource-filename.yaml":          {Data: []byte("abc")},
+			},
+			expected: []string{"resource-file-1.yaml"},
+		},
+		{
+			name: "read only resource files/dir with size > 0",
+			in: fstest.MapFS{
+				"resource-file-1.yaml":         {Data: []byte("abc")},
+				"empty-resource-filename.yaml": {},
+			},
+			expected: []string{"resource-file-1.yaml"},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got, _ := readDirForResources(tc.in)
+			if len(got) != len(tc.expected) {
+				t.Errorf("Got: %v \n", got)
+				t.Errorf("Want: %v \n", tc.expected)
+			} else {
+				gotNames := make([]string, 0)
+				for _, dir := range got {
+					gotNames = append(gotNames, dir.Name())
+				}
+				sort.Slice(gotNames, func(i, j int) bool {
+					return gotNames[i] > gotNames[j]
+				})
+				sort.Slice(tc.expected, func(i, j int) bool {
+					return tc.expected[i] > tc.expected[j]
+				})
+				if !reflect.DeepEqual(gotNames, tc.expected) {
+					t.Error("Got:")
+					for _, g := range gotNames {
+						t.Error(g)
+					}
+					t.Error("Want:")
+					for _, te := range tc.expected {
+						t.Error(te)
+					}
+				}
+			}
+		})
+	}
+}

